lib/middleware/operation.js

```diff
@@ -32,55 +36,92 @@ function findPropertyOperations ({ types, method, term, resource }) {
-      // only look for direct operation when the root resource is requested
-      operations = findClassOperations(types, method)
+    let operations = mapOperations({ api: req.hydra.api, res, method })
+      .filter(({ operation }) => operation)
+
+    if (operationMiddleware) {
+      operations = operationMiddleware(operations, { req, res })
+    }
+
+    if (!operations.length) {
+      return next()
     }
-    if (!operations) {
-      // otherwise try finding the operation by property usage
-      operations = findPropertyOperations({ types, method, term: req.hydra.term, resource: req.hydra.resource })
+
+    if (operations.length > 1) {
+      error('Multiple operations found')
+      return next(new Error('Ambiguous operation'))
     }
 
-    const [operation, ...rest] = (operations || []).toArray()
+    const [{ resource, operation }] = operations
+    req.hydra.resource = resource
+    req.hydra.operation = operation
+    return next()
+  }
+}
 
-    if (!operation) {
+async function invokeOperation (req, res, next) {
+  const { api } = req.hydra
+
+  if (!req.hydra.operation) {
+    const operationMap = mapOperations({ api, res })
+    const allowedMethods = new Set(operationMap
+      .filter(({ operation }) => operation)
+      .flatMap(({ operation }) => operation.out(ns.hydra.method).values))
+
+    if (!allowedMethods.size) {
       warn('no operations found')
-      let allowedOperations
-      if (req.hydra.term.equals(req.hydra.resource.term)) {
-        allowedOperations = findClassOperations(types)
-      } else {
-        allowedOperations = findPropertyOperations({ types, term: req.hydra.term, resource: req.hydra.resource })
+      if (operationMap.every(({ resource }) => 'property' in resource)) {
+        return next(new createError.NotFound())
       }
-
-      const allowedMethods = new Set(allowedOperations.out(ns.hydra.method).values)
-      res.setHeader('Allow', [...allowedMethods])
       return next(new createError.MethodNotAllowed())
     }
 
-    if (rest.length > 0) {
-      error('Multiple operations found')
-      return next(new Error('Ambiguous operation'))
-    }
+    res.setHeader('Allow', [...allowedMethods])
+    return next(new createError.MethodNotAllowed())
+  }
 
-    const handler = await api.loaderRegistry.load(operation.out(code.implementedBy), { basePath: api.codePath })
+  const handler = await api.loaderRegistry.load(req.hydra.operation.out(code.implementedBy), { basePath: api.codePath })
 
-    if (!handler) {
-      return next(new Error('Failed to load operation'))
-    }
+  if (!handler) {
+    return next(new Error('Failed to load operation'))
+  }
 
-    req.hydra.operation = operation
-    handler(req, res, next)
+  handler(req, res, next)
+}
+
+function factory (middleware = {}) {
+  const router = Router()
+
+  router.use(findOperation(middleware.operations))
+
+  if (middleware.resource) {
+    router.use(middleware.resource)
```

Comment:
I'm a little concerned about this. The `middleware.resource` must be aware that `req.hydra.resource` can be undefined if no resource+operation pair was indeed found.

Maybe move it inside `invokeOperation` somehow?